tarofreighter/wallet.go

```diff
@@ -0,0 +1,717 @@
+	// we'll add the relevant information to the virtual TX's input.
+	//
+	// TODO(roasbeef): still need to add family key to PrevID.
+	vPkt.Inputs = []*taropsbt.VInput{{
+		PrevID: asset.PrevID{
+			OutPoint: assetInput.AnchorPoint,
+			ID:       assetInput.Asset.ID(),
+			ScriptKey: asset.ToSerialized(
+				assetInput.Asset.ScriptKey.PubKey,
+			),
+		},
+		Anchor: taropsbt.Anchor{
+			Value:             assetInput.AnchorOutputValue,
+			PkScript:          anchorPkScript,
+			InternalKey:       internalKey.PubKey,
+			MerkleRoot:        anchorMerkleRoot,
+			Bip32Derivation:   inBip32Derivation,
+			TrBip32Derivation: inTrBip32Derivation,
+		},
+		PInput: psbt.PInput{
+			SighashType: txscript.SigHashDefault,
+		},
+	}}
+	vPkt.SetInputAsset(0, assetInput.Asset, inputProof)
+
+	// We'll validate the selected input and commitment. From this we'll
+	// gain the asset that we'll use as an input and info w.r.t if we need
+	// to use an un-spendable zero-value root.
+	inputAsset, fullValue, err := taroscript.IsValidInput(
+		assetInput.Commitment, fundDesc,
+		*vPkt.Inputs[0].Asset().ScriptKey.PubKey,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// We expect some change back, so let's create a script key to receive
+	// the change on.
+	if !fullValue {
+		senderScriptKey, err := f.cfg.KeyRing.DeriveNextKey(
+			ctx, asset.TaroKeyFamily,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		// We'll assume BIP 86 everywhere, and use the tweaked key from
+		// here on out.
+		vPkt.Outputs[0].Amount = inputAsset.Amount - fundDesc.Amount
+		vPkt.Outputs[0].ScriptKey = asset.NewScriptKeyBIP0086(
+			senderScriptKey,
+		)
+	}
+
+	// Before we can prepare output assets for our send, we need to generate
+	// a new internal key for the anchor output of the asset change output.
+	changeInternalKey, err := f.cfg.KeyRing.DeriveNextKey(
+		ctx, asset.TaroKeyFamily,
+	)
+	if err != nil {
+		return nil, err
+	}
+	vPkt.Outputs[0].SetAnchorInternalKey(
+		changeInternalKey, f.cfg.ChainParams.HDCoinType,
+	)
+
+	if err := taroscript.PrepareOutputAssets(vPkt); err != nil {
+		return nil, fmt.Errorf("unable to create split commit: %w",
+			err)
+	}
+
+	return assetInput.Commitment, nil
+}
+
+// SignVirtualPacket signs the virtual transaction of the given packet and
+// returns the input indexes that were signed (referring to the virtual
+// transaction's inputs).
+//
+// NOTE: This is part of the Wallet interface.
+func (f *AssetWallet) SignVirtualPacket(vPkt *taropsbt.VPacket) ([]uint32,
+	error) {
+
+	// Now we'll use the signer to sign all the inputs for the new taro
+	// leaves. The witness data for each input will be assigned for us.
+	signedInputs := make([]uint32, len(vPkt.Inputs))
+	for idx := range vPkt.Inputs {
+		// Before we sign the transaction, we want to make sure the
+		// inclusion proof is valid and the asset is actually committed
+		// in the anchor transaction.
+		err := verifyInclusionProof(vPkt.Inputs[idx])
```

Comment:
Probably want a TODO here to add a call to `ChainBridge` later on? To verify the anchor TX from the proof is in the chain.

Reply:
Great point! That also reminded me that we didn't check the anchor TX in the input vs. the one in the proof, which I added now. So only the "is the block in the chain" logic remains, for which I added a TODO.